steward/src/cork/proposals.rs

```diff
@@ -0,0 +1,372 @@
+
+pub async fn start_scheduled_cork_proposal_polling_thread() -> JoinHandle<()> {
+    debug!("starting cork proposal polling thread");
+    let config = APP.config();
+    let mut state = ProposalThreadState::default();
+
+    let query_period = Duration::from_secs(config.cork.proposal_poll_period);
+
+    tokio::spawn(async move {
+        let mut fail_count = 0;
+        loop {
+            tokio::time::sleep(query_period).await;
+
+            debug!("polling for new approved scheduled cork proposals");
+            state
+                .update_latest_observed_height(config.cosmos.grpc.clone())
+                .await;
+            if let Err(err) = poll_approved_cork_proposals(&mut state).await {
+                fail_count += 1;
+                error!(
+                    "failed to poll approved scheduled cork proposals {} time(s): {}",
+                    fail_count, err
+                );
+            } else {
+                fail_count = 0;
+            }
+        }
+    })
+}
+
+#[derive(Debug, Default)]
+struct ProposalThreadState {
+    proposal_query_offset: u64,
+    last_observed_height: u64,
+    retry_data: Vec<(u64, String, Vec<u8>, u64, u8)>,
+}
+
+impl ProposalThreadState {
+    async fn update_latest_observed_height(&mut self, endpoint: String) {
+        match TendermintQueryClient::connect(endpoint).await {
+            Ok(mut client) => {
+                match client.get_latest_block(GetLatestBlockRequest {}).await {
+                    Ok(r) => {
+                        let r = r.into_inner();
+                        if r.block.is_some() {
+                            let block = r.block.unwrap();
+                            if block.header.is_some() {
+                                self.last_observed_height = block.header.unwrap().height as u64;
+                            }
+                        }
+                    }
+                    Err(err) => warn!("failed to query latest block height. this could eventually result in degraded scheduled cork proposal query performance: {}", err),
+                };
+            }
+            Err(err) => warn!("failed to query latest block height. this could eventually result in degraded scheduled cork proposal query performance: {}", err),
+        };
+    }
+}
+
+async fn poll_approved_cork_proposals(state: &mut ProposalThreadState) -> Result<(), Error> {
+    let config = APP.config();
+    let mut gov_client = GovQueryClient::connect(config.cosmos.grpc.clone()).await?;
+    let mut pagination = PageRequest {
+        limit: 20,
+        offset: state.proposal_query_offset,
+        ..Default::default()
+    };
+    let mut processed_total: u64 = 0;
+
+    loop {
+        debug!("querying proposals");
+        let proposals = gov_client
+            .proposals(tonic::Request::new(gravity_bridge::gravity_proto::cosmos_sdk_proto::cosmos::gov::v1beta1::QueryProposalsRequest {
```

Comment:
Similarly here, this roundabout type import gets super ugly. If we have to keep it this way, We should probably have an import statement up top for a large portion of these prefixes.